### PR TITLE
Added .ContinueWith on Task.Delay to surpress exceptions thrown during Dispose

### PR DESCRIPTION
**Pull request by jamarino**
Fixes issue #91

While I admit it is not the most expressive way of dealing with TaskCancellationExceptions (hence the comment), it is a common way of doing it.

### CODE REVIEW

**Comment by Choc13** [comment]
@jamarino thanks for submitting this PR.

I've had a bit of a read into how Microsoft recommend handling task cancellation, specifically when creating background and child tasks. [This article](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/how-to-cancel-a-task-and-its-children) provides some good information.

From this it would appear that we should be doing a couple more things:
1. Passing the `CancellationToken` to the `Task.Run` when `pollTask` is created.
2. Calling `ThrowIfCancellationIsRequested()` in the polling loop when the token is cancelled, rather than just existing silently. 
3. Wrapping the disposal in a `try/catch` block and suppressing the `OperationCanceledException`

Personally, I don't really like this approach of throwing to cancel, but if that's that tasks are designed then we should probably stick to the rules. Would you like to make those changes and try it out?

**Comment by Choc13** [comment]
For reference, this is the bit of code I was using to validate that this indeed works. Hopefully it helps highlight the steps above a bit better too:

```csharp
private void TestingTaskCancellation()
{
    Task delay = null;

    // Function simulates the polling loop
    async Task Poll(CancellationToken ct)
    {
        while (true)
        {
            ct.ThrowIfCancellationRequested();
            Console.WriteLine("Doing some work");
            delay = Task.Delay(1000, ct);
            await delay;
        }
    }

    // This simulates the polling loop setup
    var cancel = new CancellationTokenSource();
    var poll = Poll(cancel.Token);
    var backgroundTask = Task.Run(() => poll, cancel.Token);

    // Simulating our Dispose method
    try
    {
        cancel.Cancel();
        Task.Delay(5000).Wait();
        backgroundTask.Wait(500, cancel.Token);
        cancel.Dispose();
    }
    catch (OperationCanceledException)
    {
        Console.WriteLine("Tasks shut down");
        Console.WriteLine($"Background Task is: {backgroundTask.Status}");
        Console.WriteLine($"Polling is: {poll.Status}");
        Console.WriteLine($"Delay is: {delay.Status}");
    }
}
```

**Comment by jamarino** [comment]
@Choc13 you raise some valid points, I'll have another look tonight and incorporate your suggestions.

**Comment by jamarino** [comment]
@Choc13 I've done some reading and I've settled on a solution that I am fairly happy with.

A handful of observations:

1) Yes, purposely throwing exceptions and wrapping everything in try/catch seems silly. I usually avoid it if at all possible.

2) The dispose method does not necessarily need to wait for the `PollingLoop` to stop fully. It just needs to ensure that resources are eventually released (including `_cancellationTokenSource`).
The `CancellationTokenSource` can be cleaned up without waiting for `_pollingTask` by adding a `.ContinueWith()` that calls `.Dispose()` once the task completes (ie. is cancelled).

3) I've seen some people comment that it is not necessary to dispose a `CancellationTokenSource` that has already been cancelled. Apparently cancelling does the clean up as well. If this is the case, we could settle for simply calling cancel on the token source in the dispose method.
Others argue that since `CancellationTokenSource` implements `IDisposable` you should always do your best to dispose the object when you are done with it.

4) It is generally recommended to pass in a token (not source) in the methods that require it. By referencing the source from within the methods you risk the reference changing (in this case it is read-only, so no issue) and if the source is disposed an exception is thrown when accessing `CancellationTokenSource.Token`.
I had a shot at adjusting the methods to accept a `CancellationToken` where applicable, but I ended up having issues with one of the tests (`ShouldCancelPollingTaskWhenReloading()`).

I'd love to hear what you think, and I am happy to give it another round if you have any reservations or suggestions.

**Comment by jamarino** [comment]
I had another go at it to address point (4) in my previous comment.

> It is generally recommended to pass in a token (not source) in the methods that require it.

By adding a CancellationToken as method parameter in a few places we no longer reference the CancellationTokenSource from the polling loop, and thus we can safely cancel and dispose it without waiting for the loop to stop.

This greatly simplifies the Dispose implementation and obsoletes point (2) and (3) in my previous comment. (Although I'll keep (2) in mind in future, it's a pretty neat trick).

I hope you don't mind me trying a few different approaches, I am still getting to grips with async/await and cancellation tokens.

**Comment by Choc13** [comment]
@jamarino thanks for all the investigation here and I think everything you have said makes sense. I was aware of the separation between`CancellationTokenSource` and `CancellationToken` but, as you say, because the source is immutable I didn't worry too much about referencing `cancellationToken.Token` directly. However, as you're highlighted this means the clean up is more complicated because of the reference to the disposable resource. 

I like your changes, I think they have made the code more correct and got rid of the awkward `Wait` in the `Dispose` method. I just have a few small comments on the code style to keeps things consistent with the rest of this library which I'll add now. 

Thanks.

File: test/Winton.Extensions.Configuration.Consul.Test/ConsulConfigurationProviderTests.cs [hunk 1]
```diff
@@ -87,18 +87,23 @@ private async Task ShouldCancelPollingTaskWhenReloading()
...
-                                cancellationToken = token;
+                                token.Register(() => pollingCancelled.TrySetResult(true));
                             }
                         })
-                    .Returns(
-                        pollingCancelled.Task.IsCompleted
-                            ? Task.FromCanceled<QueryResult<KVPair[]>>(pollingCancelled.Task.Result)
-                            : Task.FromResult(new QueryResult<KVPair[]> { StatusCode = HttpStatusCode.OK }));
+                    .Returns<string, QueryOptions, CancellationToken>(
+                        (_, __, token) =>
```
Comment (Choc13) [inline test/Winton.Extensions.Configuration.Consul.Test/ConsulConfigurationProviderTests.cs]: You can do away with the `return` here and do:

```
(_, __, token) => token.IsCancellationRequested ?
    ? Task.FromCanceled<QueryResult<KVPair[]>>(token)
    : Task.Delay(5).ContinueWith(t => new QueryResult<KVPair[]> { StatusCode = HttpStatusCode.OK })
```

**Comment by jamarino** [comment]
Hi @Choc13, glad you like the changes! I'll implement your suggestions tonight when I get home.

File: src/Winton.Extensions.Configuration.Consul/ConsulConfigurationProvider.cs [hunk 2]
```diff
@@ -57,20 +56,22 @@ public override void Load()
                 return;
             }
 
-            DoLoad().GetAwaiter().GetResult();
+            var ct = _cancellationTokenSource.Token;
```
[the remark this answers is not in the document]
Comment (jamarino) [reply src/Winton.Extensions.Configuration.Consul/ConsulConfigurationProvider.cs]: Done and done!

File: test/Winton.Extensions.Configuration.Consul.Test/ConsulConfigurationProviderTests.cs (hunk 1 above)
Comment (jamarino) [reply test/Winton.Extensions.Configuration.Consul.Test/ConsulConfigurationProviderTests.cs]: You've got it! 👍

**Comment by Choc13** [comment]
LGTM. Could you also open the identical PR to branch release/3.0 so I can create a patch release for these changes? I’ll put this into 3.0.1.

**Pull request merged**